File: feb-14-champagne-tower/solution.js
/**
 * @param {number} poured
 * @param {number} query_row
 * @param {number} query_glass
 * @return {number}
 */
var champagneTower = function(poured, query_row, query_glass) {
  let row = [poured];

  for (let r = 0; r < query_row; r++) {
    const next = new Array(row.length + 1).fill(0);
    for (let i = 0; i < row.length; i++) {
      const overflow = Math.max(0, row[i] - 1) / 2;
      if (overflow > 0) {
        next[i] += overflow;
        next[i + 1] += overflow;
      }
    }
    row = next;
  }

  return Math.min(1, row[query_glass]);
};

File: feb-14-champagne-tower/solution.java
class Solution {
    public double champagneTower(int poured, int query_row, int query_glass) {
        double[] row = new double[] {poured};

        for (int r = 0; r < query_row; r++) {
            double[] next = new double[row.length + 1];
            for (int i = 0; i < row.length; i++) {
                double overflow = Math.max(0.0, row[i] - 1.0) / 2.0;
                if (overflow > 0.0) {
                    next[i] += overflow;
                    next[i + 1] += overflow;
                }
            }
            row = next;
        }

        return Math.min(1.0, row[query_glass]);
    }
}
